src/Shared/NativeMethodsShared.cs

```diff
@@ -1181,19 +1236,37 @@ internal static int GetParentProcessId(int processId)
         /// Internal, optimized GetCurrentDirectory implementation that simply delegates to the native method
         /// </summary>
         /// <returns></returns>
-        internal static string GetCurrentDirectory()
+        internal unsafe static string GetCurrentDirectory()
         {
+#if FEATURE_LEGACY_GETCURRENTDIRECTORY
             if (IsWindows)
             {
-                StringBuilder sb = new StringBuilder(MAX_PATH);
-                int pathLength = GetCurrentDirectory(MAX_PATH, sb);
-
-                return pathLength > 0 ? sb.ToString() : null;
+                int bufferSize = GetCurrentDirectoryAndAssertSuccess(0, null);
+                char* buffer = stackalloc char[bufferSize];
+                int pathLength = GetCurrentDirectoryAndAssertSuccess(bufferSize, buffer);
+                return new string(buffer, startIndex: 0, length: pathLength);
             }
-
+#endif
             return Directory.GetCurrentDirectory();
         }
 
+        private unsafe static int GetCurrentDirectoryAndAssertSuccess(int nBufferLength, char* lpBuffer)
```

Comment:
I don't know if you need to explicitly say `AndAssertSuccess` in the name, since it's the default .NET error-handling strategy to throw on failure when in a method that doesn't start with `Try...`.